tuf/ngclient/updater.py

```
@@ -362,9 +363,12 @@ def _update_root_symlink(self) -> None:
         linkname = os.path.join(self._dir, "root.json")
         version = self._trusted_set.root.version
         current = os.path.join("root_history", f"{version}.root.json")
-        with contextlib.suppress(FileNotFoundError):
-            os.remove(linkname)
-        os.symlink(current, linkname)
+
+        with open(os.path.join(self._dir, current + ".lck"), "wb") as f:
+            lock_file(f)
```

A comment explaining why lock_file() is called might make sense here (maybe mention windows issues specifically for anyone who tries to repro in future)

---

I have run into some issues on Windows. When I dump filenames, it shows me something like this.
```
C:\Users\StefanBerger\AppData\Local\sigstore\sigstore-python\tuf\https%3A%2F%2Ftuf-repo-cdn.sigstore.dev\root_history\12.root.json
```
I cannot access this file. When trying to walk the directory structure it ends for me at `C:\...\Local`, so there's no `sigstore` directory that I could cd into. Then I tried to switch to locking via a single file at `%TEMP%\lck` -- it works 'better' but strangely not 100% reliable like it does on Linux.

---

Because of Windows I would need to switch to filelock library, which already is available when installing for example dev or build requirements. As it looks I may still need to either entirely switch to different file paths for locking or at least use %TEMP% file paths on Windows.

---

Without using any locking, it fails quickly on Windows. With locking (from filelock package for example) it works better but not 100% since after some time (several minutes) it again fails. Very odd.

---

hmm, let me check if the directory is guaranteed to exist at this point...

---

I think the directory always has to exist by the time `_update_root_symlink()` is called -- but I don't mind if you add a

        Path(self._dir).mkdir(exist_ok=True, parents=True)

as the first thing in `_load_root()` just to be sure.


In any case, I think this

> I cannot access this file

sounds like something we need to troubleshoot before we continue to other things.

---

> > I cannot access this file
> 
> sounds like something we need to troubleshoot before we continue to other things.

It is as if the directory structure does not exist on disk. For this reason I now have a single lock in %TEMP%/tuf.root.json.lck, which improves things a bit but is not 100% reliable, either.

Let me just push what I currently experiment with on Windows. I now run into this problem here:
```
[...]
  File "C:\Users\StefanBerger\python-tuf\tuf\ngclient\updater.py", line 145, in __init__
    self._persist_root(self._trusted_set.root.version, bootstrap)
    ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\StefanBerger\python-tuf\tuf\ngclient\updater.py", line 342, in _persist_root
    self._persist_file(str(rootdir / f"{version}.root.json"), data)
    ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\StefanBerger\python-tuf\tuf\ngclient\updater.py", line 365, in _persist_file
    raise e
  File "C:\Users\StefanBerger\python-tuf\tuf\ngclient\updater.py", line 358, in _persist_file
    os.replace(temp_file.name, filename)
    ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^

It reports:

[WinError 5] Access is denied: 'C:\\Users\\StefanBerger\\AppData\\Local\\sigstore\\sigstore-python\\tuf\\https%3A%2F%2Ftuf-repo-cdn.sigstore.dev\\tmpfi0pwqfo' -> 'C:\\Users\\StefanBerger\\AppData\\Local\\sigstore\\sigstore-python\\tuf\\https%3A%2F%2Ftuf-repo-cdn.sigstore.dev\\root_history\\12.root.json'
```

---

* using TEMP is not a solution: the reason for running this code is to create the files in the metadata directory so that has to work, otherwise the code is useless
* I'm a bit confused, can you reproduce the "missing directory" issue with code from main (as in doing whatever you did to get the failure but without your changes)? We do run tests on windows so I would expect this code to work there ... but Windows has never been an important platform so it's not impossible that there are platform bugs

Feel free to include the full error in your copy-paste (including the exception name) if you need to include another one.

---

I thought maybe filelock has a subtle bug (it's deprecated), so I moved onto fasterners package but it does **not** solve the above PermissionError problem either. I tried to protect `12.root.json` from concurrent access with 3 different locks now. I am not sure where else it could be accessed so that this PermissionError could occur. The loop solves it but it should not be necessary.

```
            lck = os.path.join(self._dir, ".root.json.lck")
            ipl = InterProcessLock(create_lockfile(lck))
            ipl.acquire()
            try:
                while True:
                    try:
                        os.replace(temp_file.name, filename)
                        break
                    except PermissionError as pe:
                        print(f"{pe}")
            finally:
                ipl.release()
```

---

another question: does the "Access is denied" only happen when you run your parallel execution test case? 

Because if it does only happen in that case, then the "Access is denied" actually means "the file is used by another process", meaning we actually need the more extensive locking process: handling just the symlink is not enough

---

> another question: does the "Access is denied" only happen when you run your parallel execution test case?

Yes. And this is the only reason I see it is failing now when I let it propagate the Exception.
> 
> Because if it does only happen in that case, then the "Access is denied" actually means "the file is used by another process", meaning we actually need the more extensive locking process: handling just the symlink is not enough

That's what I guessed as well and that's the reason I now have 3 locks. Though maybe there is still another path were this file is being used concurrently. The test is still the simple 2 liner loop... if you have an idea...

---

Thanks for testing, I will have a look at the code: I think multiple separate locks during refresh() is unappealing but maybe setting up a single lock for the whole method is ok... Still need to think of something reasonable for dwnload_target() as I'm sure that will conflict in the same way -- and in that case locking for the whole method might not be appropriate.

But (unless you solve everything in the meantime) I'll have a look with fresh eyes tomorrow and hopefully come up with something.